sdk/nodejs/runtime/closure/codePaths.ts

```diff
@@ -0,0 +1,186 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as fs from "fs";
+import * as filepath from "path";
+import * as readPackageTree from "read-package-tree";
+import * as asset from "../../asset";
+import { RunError } from "../../errors";
+
+// computeCodePaths computes the local node_module paths to include in an uploaded Lambda.
+export async function computeCodePaths(
+    extraIncludePaths?: string[],
+    extraIncludePackages?: string[],
+    extraExcludePackages?: string[]): Promise<Map<string, asset.Asset | asset.Archive>> {
+
+    // Construct the set of paths to include in the archive for upload.
+    extraIncludePaths = extraIncludePaths || [];
+    extraIncludePackages = extraIncludePackages || [];
+    extraExcludePackages = extraExcludePackages || [];
+
+    const includedPackages = new Set<string>();
+    const excludedPackages = new Set<string>();
+
+    for (const p of extraExcludePackages) {
+        excludedPackages.add(p);
+    }
+
+    for (const p of extraIncludePackages) {
+        includedPackages.add(p);
+    }
+
+    // Find folders for all packages requested by the user
+    const pathSet = await allFoldersForPackages(includedPackages, excludedPackages);
+
+    // Add all paths explicitly requested by the user
+    for (const path of extraIncludePaths) {
+        pathSet.add(path);
+    }
+
+    const codePaths: Map<string, asset.Asset | asset.Archive> = new Map();
+
+    // For each of the required paths, add the corresponding FileArchive or FileAsset to the AssetMap.
+    for (const path of pathSet.values()) {
+        // The Asset model does not support a consistent way to embed a file-or-directory into an `AssetArchive`, so
+        // we stat the path to figure out which it is and use the appropriate Asset constructor.
+        const stats = fs.lstatSync(path);
+        if (stats.isDirectory()) {
+            codePaths.set(path, new asset.FileArchive(path));
+        }
+        else {
+            codePaths.set(path, new asset.FileAsset(path));
+        }
+    }
+
+    return codePaths;
+}
+
+const lambdaRolePolicy = {
```

Comment:
Is this dead code now? We shouldn't have AWS details in `pulumi/pulumi`.

Reply:
yes.  thanks!

Reply:
Done